File: backend-C#/Controllers/ChatController.cs
using IMS.DTOs;
using IMS.DTOs.Requests;
using IMS.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace IMS.Controllers;

[ApiController]
[Route("api/chat")]
[Authorize]
public class ChatController : BaseController
{
    private readonly IChatService _chatService;

    public ChatController(IChatService chatService)
    {
        _chatService = chatService;
    }

    [HttpGet("list")]
    public async Task<ActionResult<ApiResponse>> GetChatList()
    {
        if (!CurrentUserId.HasValue)
        {
            return Error("未登录", 401);
        }

        var result = await _chatService.GetChatListAsync(CurrentUserId.Value);
        return Success(new { contacts = result.Contacts, total = result.Total }, "获取成功");
    }

    [HttpGet("history/{userId}")]
    public async Task<ActionResult<ApiResponse>> GetChatHistory(long userId, [FromQuery] ChatHistoryRequest request)
    {
        if (!CurrentUserId.HasValue)
        {
            return Error("未登录", 401);
        }

        var result = await _chatService.GetChatHistoryAsync(CurrentUserId.Value, userId, request);
        return Success(new 
        { 
            messages = result.Messages, 
            total = result.Total, 
            page = result.Page, 
            limit = result.Limit 
        }, "获取成功");
    }

    [HttpPost("send")]
    public async Task<ActionResult<ApiResponse>> SendMessage([FromBody] SendMessageRequest request)
    {
        if (!CurrentUserId.HasValue)
        {
            return Error("未登录", 401);
        }

        if (request.ReceiverId == CurrentUserId.Value)
        {
            return Error("不能给自己发送消息", 400);
        }

        try
        {
            var message = await _chatService.SendMessageAsync(CurrentUserId.Value, request);
            return Success(message, "发送成功");
        }
        catch (InvalidOperationException ex)
        {
            return Error(ex.Message, 400);
        }
    }

    [HttpPut("read/{userId}")]
    public async Task<ActionResult<ApiResponse>> MarkAsRead(long userId)
    {
        if (!CurrentUserId.HasValue)
        {
            return Error("未登录", 401);
        }

        var count = await _chatService.MarkAsReadAsync(CurrentUserId.Value, userId);
        return Success(new { markedCount = count }, "标记成功");
    }

    [HttpGet("unread")]
    public async Task<ActionResult<ApiResponse>> GetUnreadCount()
    {
        if (!CurrentUserId.HasValue)
        {
            return Error("未登录", 401);
        }

        var count = await _chatService.GetUnreadCountAsync(CurrentUserId.Value);
        return Success(new { unreadCount = count }, "获取成功");
    }

    [HttpGet("admin/messages")]
    [Authorize(Roles = "admin")]
    public async Task<ActionResult<ApiResponse>> GetAllMessages(
        [FromQuery] int page = 1, 
        [FromQuery] int limit = 50, 
        [FromQuery] string? search = null)
    {
        if (!CurrentUserId.HasValue)
        {
            return Error("未登录", 401);
        }

        var result = await _chatService.GetAllMessagesAsync(CurrentUserId.Value, page, limit, search);
        return Success(new 
        { 
            messages = result.Messages, 
            total = result.Total 
        }, "获取成功");
    }

    [HttpGet("admin/conversations")]
    [Authorize(Roles = "admin")]
    public async Task<ActionResult<ApiResponse>> GetAllConversations(
        [FromQuery] int page = 1, 
        [FromQuery] int limit = 50, 
        [FromQuery] string? search = null)
    {
        if (!CurrentUserId.HasValue)
        {
            return Error("未登录", 401);
        }

        var result = await _chatService.GetAllConversationsAsync(CurrentUserId.Value, page, limit, search);
        return Success(new 
        { 
            conversations = result.Contacts, 
            total = result.Total 
        }, "获取成功");
    }
}

File: backend-JS/src/controllers/chatController.js
const { chatService } = require('../services');
const { success, error } = require('../utils/response');

const getChatList = async (req, res) => {
  try {
    const contacts = await chatService.getChatList(req.userId);
    return success(res, { contacts, total: contacts.length }, '获取成功');
  } catch (err) {
    return error(res, 500, '获取联系人失败');
  }
};

const getChatHistory = async (req, res) => {
  try {
    const otherUserId = parseInt(req.params.userId);
    const page = parseInt(req.query.page) || 1;
    const limit = parseInt(req.query.limit) || 50;
    
    const result = await chatService.getChatHistory(req.userId, otherUserId, page, limit);
    return success(res, result, '获取成功');
  } catch (err) {
    return error(res, 500, '获取消息失败');
  }
};

const sendMessage = async (req, res) => {
  try {
    const { receiverId, content } = req.body;
    
    if (!receiverId || !content) {
      return error(res, 400, '请求参数错误');
    }
    
    const message = await chatService.sendMessage(req.userId, receiverId, content);
    return success(res, message, '发送成功');
  } catch (err) {
    return error(res, 500, '发送消息失败');
  }
};

const markAsRead = async (req, res) => {
  try {
    const otherUserId = parseInt(req.params.userId);
    await chatService.markAsRead(req.userId, otherUserId);
    return success(res, null, '已标记为已读');
  } catch (err) {
    return error(res, 500, '操作失败');
  }
};

const getUnreadCount = async (req, res) => {
  try {
    const result = await chatService.getUnreadCount(req.userId);
    return success(res, result, '获取成功');
  } catch (err) {
    return error(res, 500, '获取失败');
  }
};

const getAllMessages = async (req, res) => {
  try {
    const page = parseInt(req.query.page) || 1;
    const limit = parseInt(req.query.limit) || 20;
    
    const result = await chatService.getAllMessages(page, limit);
    return success(res, result, '获取成功');
  } catch (err) {
    console.error('[getAllMessages] Error:', err.message);
    console.error('[getAllMessages] Stack:', err.stack);
    return error(res, 500, '获取消息失败');
  }
};

const getAllConversations = async (req, res) => {
  try {
    console.log('getAllConversations called with query:', req.query);
    const page = parseInt(req.query.page) || 1;
    const limit = parseInt(req.query.limit) || 50;
    const conversations = await chatService.getAllConversations(page, limit);
    return success(res, { conversations }, '获取成功');
  } catch (err) {
    console.error('Error in getAllConversations:', err);
    return error(res, 500, '获取会话失败');
  }
};

module.exports = {
  getChatList,
  getChatHistory,
  sendMessage,
  markAsRead,
  getUnreadCount,
  getAllMessages,
  getAllConversations
};
